### PR TITLE
[TTAHUB-2546] TR goal: destroy goal when session recip removed.

### PR DESCRIPTION
**Pull request by nvms**
## Description of change

This was missed in the original TR goals PR.

### Unused on an AR
- Create a session with a recipient or two.
- Verify the goal was created.
- Remove a recipient from the session.
- Verify the entry in `Goal` and `EventReportPilotGoals` was destroyed.

### Used on an AR
- Create a session with a recipient or two.
- Verify the goal was created.
- Use the goal on an AR.
- Remove the recipient from the session.
- Verify the goal still exists because it is in use.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2546

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/hooks/sessionReportPilot.js [hunk 1]
```diff
@@ -213,6 +213,76 @@ export const createGoalsForSessionRecipientsIfNecessary = async (sequelize, sess
...
+    let event;
+    let nextSessionRecipients;
+
+    if (sessionReport && sessionReport.data) {
+      const data = (typeof sessionReport.data.val === 'string')
+        ? JSON.parse(sessionReport.data.val)
+        : sessionReport.data;
+
+      event = data.event;
+      nextSessionRecipients = data.recipients;
```
Comment (AdamAdHocTeam) [inline src/models/hooks/sessionReportPilot.js:227]: Why call it 'next' session recipients?

File: src/models/hooks/sessionReportPilot.js [hunk 2]
```diff
@@ -213,6 +213,76 @@ export const createGoalsForSessionRecipientsIfNecessary = async (sequelize, sess
...
+      where: { eventId: event.id, sessionId: sessionReport.id },
+      transaction: options.transaction,
+      raw: true,
+    });
+
+    if (!recipients || !recipients.length) return;
+
+    // Using the EventReportPilotGoal table, find all goals that were created for this recipient (grantId).
+    // eslint-disable-next-line no-restricted-syntax
+    for await (const entry of recipients) {
```
Comment (AdamAdHocTeam) [inline src/models/hooks/sessionReportPilot.js:244]: nitpick entry = recipient

File: src/models/hooks/sessionReportPilot.js [hunk 3]
```diff
@@ -213,6 +213,76 @@ export const createGoalsForSessionRecipientsIfNecessary = async (sequelize, sess
...
+      if (!nextSessionRecipients.includes(entry.grantId)) {
+        // A recipient was removed.
+
+        // If the goal is referenced in an ARG, then it's in use and cannot be deleted.
+        // This should _never_ be possible, because this hook happens when a session is edited.
+        // If the session is edited, it's not complete, and therefore the goal should not be in use.
+        const args = await sequelize.models.ActivityReportGoal.findAll({
```
Comment (AdamAdHocTeam) [inline src/models/hooks/sessionReportPilot.js:251]: Nitpick args = arguments, maybe activityReportGoals.

File: src/models/hooks/sessionReportPilot.js [hunk 4]
```diff
@@ -213,6 +213,76 @@ export const createGoalsForSessionRecipientsIfNecessary = async (sequelize, sess
...
+        });
+
+        if (!args || !args.length) {
+          // Destroy the Goal and the EventReportPilotGoal entry.
+          await sequelize.models.Goal.destroy({
+            where: { id: entry.goalId },
+            transaction: options.transaction,
+          });
+
+          await sequelize.models.EventReportPilotGoal.destroy({
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/models/hooks/sessionReportPilot.js:264]: Ah wait for some reason was thinking this was ondestory.

[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/models/hooks/sessionReportPilot.js:264]: @nvms can you double check this

File: src/models/hooks/sessionReportPilot.js (hunk 2 above)
Comment (nvms) [reply src/models/hooks/sessionReportPilot.js:244]: entry is a row in `EventReportPilotGoal`, which looks like `{ eventId, goalId, recipientId, sessionId }` - so not exactly a recipient. 

I struggled with naming this variable for .5s before telling myself "it doesn't matter".

I can rename it to `recipient` if you still think it's better?

File: src/models/hooks/sessionReportPilot.js (hunk 3 above)
Comment (nvms) [reply src/models/hooks/sessionReportPilot.js:251]: Yeah.. it's activity report goals. I can make that more clear by renaming it to `activityReportGoals`, lol.

File: src/models/hooks/sessionReportPilot.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nvms) [reply src/models/hooks/sessionReportPilot.js:264]: If we are destroying the goal, then yes, we also need to destroy the reference to it in this join table.

File: src/models/hooks/sessionReportPilot.js (hunk 1 above)
Comment (nvms) [reply src/models/hooks/sessionReportPilot.js:227]: Because it is the new value of recipients being saved.

Comment (nvms) [reply src/models/hooks/sessionReportPilot.js:227]: I'm open to suggestions for variable names if you think it needs to be more clear

File: src/models/hooks/sessionReportPilot.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply src/models/hooks/sessionReportPilot.js:264]: My bad confusing tables was thinking it was SessionReportPilot. Not EventReportPilot duh.

**Review by AdamAdHocTeam — APPROVED** [review]

**Pull request merged**